lib/runtime/storage/trie.go

```diff
@@ -174,40 +169,35 @@ func (t *TrieState) Delete(key []byte) (err error) {
 
 	if currentTx := t.getCurrentTransaction(); currentTx != nil {
 		t.getCurrentTransaction().delete(string(key))
-
-	} else {
-		err := t.state.Delete(key)
-		if err != nil {
-			return err
-		}
-		t.removeMainTrieSortedKey(string(key))
+		return nil
 	}
 
-	return nil
+	return t.state.Delete(key)
 }
 
 // NextKey returns the next key in the trie in lexicographical order. If it does not exist, it returns nil.
 func (t *TrieState) NextKey(key []byte) []byte {
 	t.mtx.RLock()
 	defer t.mtx.RUnlock()
 
+	nextKey := t.state.NextKey(key)
 	if currentTx := t.getCurrentTransaction(); currentTx != nil {
-		allSortedKeys := append(t.sortedKeys, currentTx.sortedKeys...)
 		// Find key position
-		pos, found := slices.BinarySearch(allSortedKeys, string(key))
+		pos, found := slices.BinarySearch(currentTx.sortedKeys, string(key))
 		if found {
 			pos = pos + 1
 		}
 
 		// Get next key based on that position
-		if pos < len(allSortedKeys) {
-			k := allSortedKeys[pos]
-			return []byte(k)
+		if pos < len(currentTx.sortedKeys) {
+			currentTxNextKey := currentTx.sortedKeys[pos]
+			if currentTxNextKey < string(nextKey) {
```

Comment:
If our trie is empty and we start with transactions `nextKey` is nil and we have to return the nextKey from tx

```suggestion
			if nextKey == nil || currentTxNextKey < string(nextKey) {
```